ddl/table.go

```diff
@@ -738,12 +739,37 @@ func onUpdateFlashReplicaStatus(t *meta.Meta, job *model.Job) (ver int64, _ erro
 	if err != nil {
 		return ver, errors.Trace(err)
 	}
-	if tblInfo.TiFlashReplica == nil || (tblInfo.TiFlashReplica.Available == available) {
+	if tblInfo.TiFlashReplica == nil || (tblInfo.TiFlashReplica.Available == available) ||
+		(tblInfo.ID != physicalID && available == tblInfo.TiFlashReplica.IsPartitionAvailable(physicalID)) {
 		return ver, nil
 	}
 
-	if tblInfo.TiFlashReplica != nil {
+	if tblInfo.ID == physicalID {
 		tblInfo.TiFlashReplica.Available = available
+	} else if pi := tblInfo.GetPartitionInfo(); pi != nil {
+		// Partition replica become available.
+		if available {
+			allAvailable := true
+			for _, p := range pi.Definitions {
+				if p.ID == physicalID {
+					tblInfo.TiFlashReplica.AvailablePartitionIDs = append(tblInfo.TiFlashReplica.AvailablePartitionIDs, physicalID)
+				}
+				allAvailable = allAvailable && tblInfo.TiFlashReplica.IsPartitionAvailable(p.ID)
+			}
+			tblInfo.TiFlashReplica.Available = allAvailable
+		} else {
+			// Partition replica become unavailable.
+			for i, id := range tblInfo.TiFlashReplica.AvailablePartitionIDs {
+				if id == physicalID {
+					newIDs := make([]int64, 0, len(tblInfo.TiFlashReplica.AvailablePartitionIDs)-1)
+					newIDs = append(newIDs, tblInfo.TiFlashReplica.AvailablePartitionIDs[:i]...)
+					newIDs = append(newIDs, tblInfo.TiFlashReplica.AvailablePartitionIDs[i+1:]...)
+					tblInfo.TiFlashReplica.AvailablePartitionIDs = newIDs
```

Comment:
Is it necessary to copy a new `AvailablePartitionIDs`?

Reply:
great, done.